#!/usr/bin/env bash
# shellcheck disable=SC1091
# shellcheck disable=SC2153
# This script is used to run the job on the NSCC cluster.

#PBS -N merged-BRCA-bindetect-repeat
#PBS -l select=1:ncpus=64:mem=250GB
#PBS -l walltime=4:00:00
#PBS -j oe
#PBS -P personal
#PBS -q normal

# load conda environment
module load miniconda3/py38_4.8.3

conda activate /home/users/ntu/suffiazi/apps/mambaforge/envs/tobias-footprinting

#id_file should be a two-column file with the first column being the dataset ID and the second column being the dataset name
dataset_file=$ID_FILE
run_num=$RUN
run_type=$TYPES

# Initialize the associative array
declare -A datasets

# Read the two-column file into the associative array
while read -r key value; do
    datasets["$key"]=$value
done < "$dataset_file"

# Initialize a new array to store the formatted key-value pairs
formatted_datalist=()

# Loop through the keys of the associative array and format the key-value pairs
for key in "${!datasets[@]}"; do
    formatted_datalist+=("${key}_${datasets[$key]}")
done

echo "Datasets to be processed: ${formatted_datalist[*]}"

# Initialize an array to hold filepaths
filepaths=()

# Loop through datasets and construct filepaths
for dataset in "${formatted_datalist[@]}"; do
    filepaths+=("/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}-BRCA/${dataset}/${dataset}_merged_footprints.bw")
done

echo "Initializing TOBIAS BinDetect run..."

# call footprints
if TOBIAS BINDetect --motifs /home/users/ntu/suffiazi/scripts/footprinting-workflow-scripts/motif-guided/joined_filt_combined_motifs.jaspar --signals "${filepaths[@]}" --genome /home/users/ntu/suffiazi/genomes/GRCh38_no_alt_GCA_000001405.15.fasta --peaks /home/users/ntu/suffiazi/scratch/inputs/peak_files/brca_subtype_merged/BRCA-master-peakset-natsorted.bed --outdir "/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}-BRCA/bindetect_outputs" --skip-excel --cond_names "${formatted_datalist[@]}" --cores 64 >> "/home/users/ntu/suffiazi/scripts/footprinting-workflow-scripts/logfiles/BRCA-merged-differential-footprinting-${run_num}-${run_type}.log"; then
    echo "Successfully ran TOBIAS footprinting for TCGA-BRCA merged data."
else
    echo "ERROR: Failed to run TOBIAS footprinting for TCGA-BRCA merged data due to an error. Please see the log file for more details."
    exit 1
fi

echo "All done! Check the log files for any errors and the outputs directory for the results."
