#!/usr/bin/env bash
# shellcheck disable=SC1091
# shellcheck disable=SC2153
# This script is used to run the job on the NSCC cluster.
# for master peakset: use 1800GB memory and 64 cores for 5h; for union peakset: use 300GB memory and 64 cores for 1h
# UPDATE: for master peakset v3 and v4: use 800GB memory and 64 cores for 5h

#PBS -N tobias-rerun2
#PBS -l select=1:ncpus=64:mem=420GB
#PBS -l walltime=4:00:00
#PBS -j oe
#PBS -P 12003580
#PBS -q normal

eval "$(conda shell.bash hook)"
conda activate /home/users/ntu/suffiazi/apps/mambaforge/envs/tobias-footprinting

dataset=$ANALYSIS_ID
sample_name=$SAMPLE
experiment_type=$EXP_TYPE
run_num=$RUN
cores=$CORE
masterpeaks=$PEAKS
bam=$FILE
run_type=$TYPES

echo "Commencing run number ${run_num} for ${dataset}_${sample_name} with ${cores} cores..."

# check if the experiment type is DNase-seq
if [[ "$experiment_type" == "dnaseseq" ]]; then
    # run bias correction
    echo "Entering DNAse-seq mode"
    echo "Running bias correction for ${dataset}_${sample_name}"
    if TOBIAS ATACorrect -b "$bam" -g /home/users/ntu/suffiazi/genomes/GRCh38_no_alt_GCA_000001405.15.fasta -p "${masterpeaks}" --blacklist /home/users/ntu/suffiazi/genomes/hg38-blacklist.v2.bed --read_shift 0 0 --k_flank 6 --outdir "/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}/${dataset}/${sample_name}" --prefix "${dataset}_${sample_name}" --cores "${cores}" >> "/home/users/ntu/suffiazi/scripts/footprinting-workflow-scripts/logfiles/${dataset}_${sample_name}_${experiment_type}-${cores}cores-aspire-${run_num}-${run_type}.log"; then
        echo "Bias correction has been completed for ${dataset}_${sample_name}. Proceeding with footprinting."
    else
        echo "ERROR: Bias correction has failed for ${dataset}_${sample_name}. Please check the log file for more details."
        exit 1
    fi
elif [[ "$experiment_type" == "atacseq" ]]; then
    # run bias correction
    echo "Entering ATAC-seq mode"
    echo "Running bias correction for ${dataset}_${sample_name}"
    if TOBIAS ATACorrect -b "$bam" -g /home/users/ntu/suffiazi/genomes/GRCh38_no_alt_GCA_000001405.15.fasta -p "${masterpeaks}" --blacklist /home/users/ntu/suffiazi/genomes/hg38-blacklist.v2.bed --outdir "/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}/${dataset}/${sample_name}" --prefix "${dataset}_${sample_name}" --cores "${cores}" >> "/home/users/ntu/suffiazi/scripts/footprinting-workflow-scripts/logfiles/${dataset}_${sample_name}_${experiment_type}-${cores}cores-aspire-${run_num}-${run_type}.log"; then
        echo "Bias correction has been completed for ${dataset}_${sample_name}. Proceeding with footprinting."
    else
        echo "ERROR: Bias correction has failed for ${dataset}_${sample_name}. Please check the log file for more details."
        exit 1
    fi
fi

FILELIMIT=$(ulimit -n)
echo "Current file descriptor limit: " "$FILELIMIT"

# run TOBIAS footprinting
# score bigwigs
echo "Scoring bigwigs for ${dataset}_${sample_name} and calling footprints..."
if TOBIAS ScoreBigwig --signal "/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}/${dataset}/${sample_name}/${dataset}_${sample_name}_corrected.bw" --regions "${masterpeaks}" --output "/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}/${dataset}/${sample_name}/${dataset}_${sample_name}_footprints.bw" --cores "${cores}" >> "/home/users/ntu/suffiazi/scripts/footprinting-workflow-scripts/logfiles/${dataset}_${sample_name}_${experiment_type}-${cores}cores-aspire-${run_num}-${run_type}.log" && \
# call footprints
TOBIAS BINDetect --motifs /home/users/ntu/suffiazi/scripts/footprinting-workflow-scripts/motif-guided/joined_filt_combined_motifs.jaspar --signals "/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}/${dataset}/${sample_name}/${dataset}_${sample_name}_footprints.bw" --genome /home/users/ntu/suffiazi/genomes/GRCh38_no_alt_GCA_000001405.15.fasta --peaks "${masterpeaks}" --outdir "/home/users/ntu/suffiazi/scratch/outputs/tobias-outs/RUN-${run_num}-${run_type}/${dataset}/${sample_name}/bindetect_outputs" --skip-excel --cores "${cores}" >> "/home/users/ntu/suffiazi/scripts/footprinting-workflow-scripts/logfiles/${dataset}_${sample_name}_${experiment_type}-${cores}cores-aspire-${run_num}-${run_type}.log"; then
    echo "Successfully ran TOBIAS footprinting for ${dataset}_${sample_name}."
    # run post-processing module
    echo "Running post-processing module for ${dataset}_${sample_name}..."
    
else
    echo "ERROR: Failed to run TOBIAS footprinting for ${dataset}_${sample_name} due to an error. Please see the log file for more details."
    exit 1
fi

echo "All done! Check the log files for any errors and the outputs directory for the results."
